### PR TITLE
Add milliseconds to report when test item started and finished

### LINKED ISSUE #3311: Minimal unit of DateTime in the report when Test was started/ended
Now I see in the report:
```xml
start-time="2019-07-04 22:32:33Z"
```

So minimal unit of time is `seconds`. It would be nice to see here more concrete date/time with milliseconds.

### PR DESCRIPTION
**Pull request by nvborisenko**
This PR fixes #3311 

Changes:
- added `o` datetime format for xml attributes
- adjusted unit tests

### CODE REVIEW

File: src/NUnitFramework/tests/Internal/Results/TestResultGeneralTests.cs [hunk 1]
```diff
@@ -160,8 +160,8 @@ public void TestResultXml_Duration()
         {
             TNode testNode = _testResult.ToXml(true);
 
-            Assert.AreEqual(EXPECTED_START.ToString("u"), testNode.Attributes["start-time"]);
-            Assert.AreEqual(EXPECTED_END.ToString("u"), testNode.Attributes["end-time"]);
+            Assert.AreEqual(EXPECTED_START, DateTime.Parse(testNode.Attributes["start-time"], CultureInfo.InvariantCulture).ToUniversalTime());
+            Assert.AreEqual(EXPECTED_END, DateTime.Parse(testNode.Attributes["end-time"], CultureInfo.InvariantCulture).ToUniversalTime());
```
Comment (jnm2) [inline src/NUnitFramework/tests/Internal/Results/TestResultGeneralTests.cs]: For our tests, I think the first way was better. Wouldn't these tests pass even if you hadn't made the change in TestResult.cs?

Comment (nvborisenko) [reply src/NUnitFramework/tests/Internal/Results/TestResultGeneralTests.cs]: I returned back everything as was, just changed `u` to `o`. Tests are passing. Also didn't use `InvariantCulture` when converting DateTime to String, seems `o` formatter is already InvariantCulture.

Comment (jnm2) [reply src/NUnitFramework/tests/Internal/Results/TestResultGeneralTests.cs]: Oh, nice to know!

**Comment by nvborisenko** [comment]
As for me I am not sure whether nunitlite should use `o` format for datetimes. As I understand correctly output in nunitlite is more for humans, while xml output for fully detailed report.

Before: `Start time: 2014-12-02 12:34:56Z`
After: `Start time: 2014-12-02T12:34:56.7890000Z`

I think after is better than before, this is more clear for humans.

Or we can improve datetime format to something like `2014-12-02 12:34:56.789` (without `T` and `Z`, use ### for milliseconds part). This variant includes milliseconds and is still clear for humans and enough detailed for xml report.

**Comment by mikkelbu** [comment]
@nvborisenko Sorry, if I was unclear. I only meant that we should correct the XML part of NUnitLite. That was the reason why I mentioned `NUnit3XmlOutputWriter.cs`. I don't think we should change what is written to the user via Console.Out.

**Comment by nvborisenko** [comment]
Reverted back changes related to console output, left changes only in `nunitlite\OutputWriters\NUnit3XmlOutputWriter.cs`.

**Comment by jnm2** [comment]
@nvborisenko Would it be easy for you to add tests on the NUnitLite change as well?

**Comment by nvborisenko** [comment]
I have added tests based on existing for `NUnit2OutputWriter`. I just mage a copy of them, so guys please review it carefully.

I am able to change or add new tests according your review, I have no enough knowledge what is expected in output.

This is xml output what I see to make review easier:
```xml
<?xml version="1.0"?>
<test-run id="2" name="MockTestFixture" fullname="NUnit.Tests.Assemblies.MockTestFixture" testcasecount="10" result="Failed" start-time="2019-09-25T19:07:18.2627674Z" end-time="2019-09-25T19:07:18.4967680Z" duration="0.233861" total="9" passed="2" failed="4" inconclusive="1" skipped="2" asserts="0" random-seed="1801742472">
  <command-line><![CDATA["C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\IDE\Extensions\TestPlatform\Extensions\..//testhost.x86.exe"  --port 54191 --endpoint 127.0.0.1:054191 --role client --parentprocessid 9320 --telemetryoptedin true]]></command-line>
  <filter/>
  <test-suite type="TestFixture" id="0-1454" name="MockTestFixture" fullname="NUnit.Tests.Assemblies.MockTestFixture" classname="NUnit.Tests.Assemblies.MockTestFixture" runstate="Runnable" testcasecount="10" result="Failed" site="Child" start-time="2019-09-25T19:07:18.2627674Z" end-time="2019-09-25T19:07:18.4967680Z" duration="0.233861" total="9" passed="2" failed="4" warnings="1" inconclusive="1" skipped="2" asserts="0">
    <environment framework-version="3.13.0.0" clr-version="4.0.30319.42000" os-version="Microsoft Windows NT 10.0.18362.0" platform="Win32NT" cwd="c:\Projects\nunit" machine-name="DESKTOP-SM1DOE9" user="nikol" user-domain="DESKTOP-SM1DOE9" culture="en-US" uiculture="en-US" os-architecture="x86"/>
    <properties>
      <property name="Description" value="Fake Test Fixture"/>
      <property name="Category" value="FixtureCategory"/>
    </properties>
    <failure>
      <message><![CDATA[One or more child tests had errors]]></message>
    </failure>
    <test-case id="0-1455" name="TestWithDescription" fullname="NUnit.Tests.Assemblies.MockTestFixture.TestWithDescription" methodname="TestWithDescription" classname="NUnit.Tests.Assemblies.MockTestFixture" runstate="Runnable" seed="1859679105" result="Passed" start-time="2019-09-25T19:07:18.2627674Z" end-time="2019-09-25T19:07:18.2657662Z" duration="0.002342" asserts="0">
      <properties>
        <property name="Description" value="Mock Test #1"/>
        <property name="Category" value="MockCategory"/>
        <property name="Severity" value="Critical"/>
      </properties>
    </test-case>
    <test-case id="0-1456" name="NonPublicTest" fullname="NUnit.Tests.Assemblies.MockTestFixture.NonPublicTest" methodname="NonPublicTest" classname="NUnit.Tests.Assemblies.MockTestFixture" runstate="NotRunnable" seed="65333886" result="Failed" label="Invalid" start-time="2019-09-25T19:07:18.2667661Z" end-time="2019-09-25T19:07:18.2667661Z" duration="0.000317" asserts="0">
      <properties>
        <property name="_SKIPREASON" value="Method is not public"/>
      </properties>
      <failure>
        <message><![CDATA[Method is not public]]></message>
      </failure>
    </test-case>
    <test-case id="0-1457" name="FailingTest" fullname="NUnit.Tests.Assemblies.MockTestFixture.FailingTest" methodname="FailingTest" classname="NUnit.Tests.Assemblies.MockTestFixture" runstate="Runnable" seed="927659394" result="Failed" start-time="2019-09-25T19:07:18.2677662Z" end-time="2019-09-25T19:07:18.4087677Z" duration="0.141415" asserts="0">
      <failure>
        <message><![CDATA[Intentional failure]]></message>
        <stack-trace><![CDATA[   at NUnit.Tests.Assemblies.MockTestFixture.FailingTest() in c:\Projects\nunit\src\NUnitFramework\mock-assembly\MockAssembly.cs:line 150
]]></stack-trace>
      </failure>
      <assertions>
        <assertion result="Failed">
          <message><![CDATA[Intentional failure]]></message>
          <stack-trace><![CDATA[   at NUnit.Tests.Assemblies.MockTestFixture.FailingTest() in c:\Projects\nunit\src\NUnitFramework\mock-assembly\MockAssembly.cs:line 150
]]></stack-trace>
        </assertion>
      </assertions>
    </test-case>
    <test-case id="0-1458" name="WarningTest" fullname="NUnit.Tests.Assemblies.MockTestFixture.WarningTest" methodname="WarningTest" classname="NUnit.Tests.Assemblies.MockTestFixture" runstate="Runnable" seed="1091845094" result="Warning" start-time="2019-09-25T19:07:18.4087677Z" end-time="2019-09-25T19:07:18.4097669Z" duration="0.000645" asserts="0">
      <reason>
        <message><![CDATA[Warning Message]]></message>
      </reason>
      <assertions>
        <assertion result="Warning">
          <message><![CDATA[Warning Message]]></message>
          <stack-trace><![CDATA[   at NUnit.Tests.Assemblies.MockTestFixture.WarningTest() in c:\Projects\nunit\src\NUnitFramework\mock-assembly\MockAssembly.cs:line 156
]]></stack-trace>
        </assertion>
      </assertions>
    </test-case>
    <test-case id="0-1459" name="IgnoreTest" fullname="NUnit.Tests.Assemblies.MockTestFixture.IgnoreTest" methodname="IgnoreTest" classname="NUnit.Tests.Assemblies.MockTestFixture" runstate="Ignored" seed="1814094590" result="Skipped" label="Ignored" start-time="2019-09-25T19:07:18.4097669Z" end-time="2019-09-25T19:07:18.4097669Z" duration="0.000005" asserts="0">
      <properties>
        <property name="_SKIPREASON" value="Ignore Message"/>
      </properties>
      <reason>
        <message><![CDATA[Ignore Message]]></message>
      </reason>
    </test-case>
    <test-case id="0-1460" name="ExplicitTest" fullname="NUnit.Tests.Assemblies.MockTestFixture.ExplicitTest" methodname="ExplicitTest" classname="NUnit.Tests.Assemblies.MockTestFixture" runstate="Explicit" seed="771945203" result="Skipped" label="Explicit" start-time="2019-09-25T19:07:18.4097669Z" end-time="2019-09-25T19:07:18.4097669Z" duration="0.000147" asserts="0"/>
    <test-case id="0-1461" name="NotRunnableTest" fullname="NUnit.Tests.Assemblies.MockTestFixture.NotRunnableTest" methodname="NotRunnableTest" classname="NUnit.Tests.Assemblies.MockTestFixture" runstate="NotRunnable" seed="2078042010" result="Failed" label="Invalid" start-time="2019-09-25T19:07:18.4097669Z" end-time="2019-09-25T19:07:18.4097669Z" duration="0.000002" asserts="0">
      <properties>
        <property name="_SKIPREASON" value="No arguments were provided"/>
      </properties>
      <failure>
        <message><![CDATA[No arguments were provided]]></message>
      </failure>
    </test-case>
    <test-case id="0-1462" name="InconclusiveTest" fullname="NUnit.Tests.Assemblies.MockTestFixture.InconclusiveTest" methodname="InconclusiveTest" classname="NUnit.Tests.Assemblies.MockTestFixture" runstate="Runnable" seed="258148165" result="Inconclusive" start-time="2019-09-25T19:07:18.4097669Z" end-time="2019-09-25T19:07:18.4527959Z" duration="0.043079" asserts="0">
      <reason>
        <message><![CDATA[No valid data]]></message>
      </reason>
    </test-case>
    <test-case id="0-1463" name="DisplayRunParameters" fullname="NUnit.Tests.Assemblies.MockTestFixture.DisplayRunParameters" methodname="DisplayRunParameters" classname="NUnit.Tests.Assemblies.MockTestFixture" runstate="Runnable" seed="161461173" result="Passed" start-time="2019-09-25T19:07:18.4527959Z" end-time="2019-09-25T19:07:18.4527959Z" duration="0.000280" asserts="0"/>
    <test-case id="0-1464" name="TestWithException" fullname="NUnit.Tests.Assemblies.MockTestFixture.TestWithException" methodname="TestWithException" classname="NUnit.Tests.Assemblies.MockTestFixture" runstate="Runnable" seed="1888992342" result="Failed" label="Error" start-time="2019-09-25T19:07:18.4527959Z" end-time="2019-09-25T19:07:18.4957673Z" duration="0.042903" asserts="0">
      <failure>
        <message><![CDATA[System.Exception : Intentional Exception]]></message>
        <stack-trace><![CDATA[   at NUnit.Tests.Assemblies.MockTestFixture.MethodThrowsException() in c:\Projects\nunit\src\NUnitFramework\mock-assembly\MockAssembly.cs:line 189
   at NUnit.Tests.Assemblies.MockTestFixture.TestWithException() in c:\Projects\nunit\src\NUnitFramework\mock-assembly\MockAssembly.cs:line 184]]></stack-trace>
      </failure>
    </test-case>
  </test-suite>
</test-run>
```

File: src/NUnitFramework/nunitlite.tests/NUnit3XmlOutputWriterTests.cs [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+            var endTimeString = RequiredAttribute(suiteNode, "end-time");
+
+            var success = DateTime.TryParse(endTimeString, out _);
+            Assert.That(success, "{0} is an invalid value for end time", endTimeString);
+        }
+
+        [Test]
+        public void TestSuite_ResultIsFailure()
+        {
+        }
```
Comment (jnm2) [inline src/NUnitFramework/nunitlite.tests/NUnit3XmlOutputWriterTests.cs]: (Empty test)

**Comment by jnm2** [comment]
> I have added tests based on existing for `NUnit2OutputWriter`. I just mage a copy of them, so guys please review it carefully.
> 
> I am able to change or add new tests according your review, I have no enough knowledge what is expected in output.

This is fantastic, thank you! So long as the new tests pass, they very helpfully describe the current state of things. If we wanted to change anything, we'd do it in a separate PR.

I'm looking specifically for a test that fails if you change the start or end formats back to `u`, just like the tests in `TestResultGeneralTests.cs`.

File: src/NUnitFramework/nunitlite.tests/NUnit3XmlOutputWriterTests.cs (hunk 2 above)
Comment (nvborisenko) [reply src/NUnitFramework/nunitlite.tests/NUnit3XmlOutputWriterTests.cs]: Removed.

**Comment by nvborisenko** [comment]
Added verification for DateTime to be exact `o` format. Using
```csharp
DateTime.TryParseExact(endTimeString, "o", CultureInfo.InvariantCulture, DateTimeStyles.None, out _)
```
in test to achieve it.

**Comment by nvborisenko** [comment]
Comparison by `string` is better.

The issue why I couldn't apply this comparison is that tested object wasn't mocked, this is real test execution result.

```csharp
ITestResult result = NUnit.TestUtilities.TestBuilder.RunTestFixture(typeof(MockTestFixture));
```

`result` object is object under test. To verify start-time and end-time we should expect some specific values, but we don't know these expected values.

**Comment by jnm2** [comment]
Ah right, thanks for pointing that out!

File: src/NUnitFramework/nunitlite.tests/NUnit3XmlOutputWriterTests.cs [hunk 3]
```diff
@@ -0,0 +1,224 @@
+// ***********************************************************************
+// Copyright (c) 2011 Charlie Poole, Rob Prouse
```
Comment (mikkelbu) [inline src/NUnitFramework/nunitlite.tests/NUnit3XmlOutputWriterTests.cs]: Also LGTM, but I have a final nit-pick. Can we change the date to 2019.

**Pull request merged**